[CI] Fix failures in Windows/Linux CI Formatting jobs by force killing all dotnet processess

4creators opened the pull request:
This PR fixes one class of CI jobs bugs: enables cleaning of CI workspace for Windows Formatting and Ubuntu Formatting jobs.

Modified python format.py script uses shell commands to kill all processess with dotnet/dotnet.exe name before
attempting the final cleanup of the CI workspace

In addition .gitignore was updated to exclude directories created during formatting job run.

4creators commented:
cc @BruceForstall

4creators commented on tests/scripts/format.py:
```diff
@@ -85,13 +85,13 @@ def main(argv):
         dotnetcliFilename = os.path.join(dotnetcliPath, 'dotnetcli-jitutils.tar.gz')
     elif platform == 'Windows_NT':
-        dotnetcliUrl = "https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.300/dotnet-sdk-2.1.300-win-x64.zip"
+        dotnetcliUrl = "https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.402/dotnet-sdk-2.1.402-win-x64.zip"
```
Updated SDK used for jobs to latest without security issues

BruceForstall replied:
nice.

4creators commented:
At the end of the execution there are still running dotnet processes:

https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_windows_nt_formatting_prtest/22282/console
```
Running: clang-format  -style=file D:\j\workspace\x64_windows_n---616f30f2\src\jit\optimizer.cpp
Shutting down MSBuild server...
Shutting down VB/C# compiler server...
MSBuild server shut down successfully.
VB/C# compiler server shut down successfully.
Stop dotnet.exe execution.
SUCCESS: The process with PID 5060 has been terminated.      <<<--- dotnet.exe process running at the end of CI job
Deleting D:\j\workspace\x64_windows_n---616f30f2\jitutils
Deleting D:\j\workspace\x64_windows_n---616f30f2\dj
Deleting D:\j\workspace\x64_windows_n---616f30f2\bootstrap.cmd
```

BruceForstall commented on tests/scripts/format.py:
```diff
@@ -227,6 +227,13 @@ def main(argv):
+    if platform == 'Linux' or platform == 'OSX':
+        subprocess.call(['killall', '-SIGTERM', '-qw', dotnet])
+    elif platform == 'Windows_NT':
+        utilpath = os.path.join(coreclr, 'tests\\scripts\\utilities.cmd')
```
"utilities.cmd" seems too generic. Why not "kill-all.cmd" or "kill-all-named-processes.cmd"?

BruceForstall commented on tests/scripts/format.py:
```diff
@@ -189,8 +189,8 @@ def main(argv):
-    builds = ["Checked", "Debug", "Release"]
-    projects = ["dll", "standalone", "crossgen"]
+    builds = ["Release"]
+    projects = ["standalone"]
```
Can you explain why you change the set of builds/projects?

BruceForstall commented on .gitignore:
```diff
@@ -37,6 +37,10 @@ msbuild.log
 *.sln.ide
 *.ide/
 
+# jitutils
```
I don't think we want this. These directories should get deleted doing `git clean -dxf`, for example. I don't know why you would want them to stick around. They're only created during the jitformat job, right?

4creators commented:
Ubuntu arm failure is unrelated to the changes:
https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/arm_cross_checked_ubuntu_no_tiered_compilation_innerloop_tst_prtest/1542/
JIT_Performance._CodeQuality_V8_Richards_Richards_Richards_._CodeQuality_V8_Richards_Richards_Richards_sh [FAIL]

@BruceForstall 
Addressed code review feedback. PTAL

4creators commented:
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test

4creators commented:
Exactly the same problems pop up in release/2.1.xx and release/2.2 branches, perhaps, this fix should be ported to those branches as well.

https://ci.dot.net/job/dotnet_coreclr/job/release_2.2/job/x64_windows_nt_formatting_prtest/146/
https://ci.dot.net/job/dotnet_coreclr/job/release_2.1/job/x64_windows_nt_formatting_prtest/381/

4creators commented:
@jashook Ping PTAL. This PR removes one group of persistent CI errors.

AndyAyersMS commented:
Do we know which `dotnet` process is still running? 

If we're asking server processes to exit and they don't we should be reporting this to the right folks.

4creators commented:
> Do we know which dotnet process is still running?

@AndyAyersMS 

Server processes are exiting correctly. The problem is with last process which kills server processes and other `dotnet` processes used during jitutils build. I have started working on fixing jitutils build process and there is PR  https://github.com/dotnet/jitutils/pull/167 which fixes some but not all problems.

AndyAyersMS commented:
If I understand python correctly then
```python
    # shutdown the dotnet build servers before cleaning things up
    proc = subprocess.Popen(["dotnet", "build-server", "shutdown"], env=my_env)
    output,error = proc.communicate()

    if os.path.isdir(jitUtilsPath):
        print("Deleting " + jitUtilsPath)
        shutil.rmtree(jitUtilsPath, onerror=del_rw)
```
should wait for that last `dotnet` to exit before it moves on to try and delete things. So how is it there is a dotnet still running at this point?

jashook commented:
It is possible that the command is exiting with some sort of error.

AndyAyersMS commented:
I'm not objecting to your change.

But I have heard discussions of possible tool server process issues and wonder if something like that is lurking here. If so it would be nice to get that sorted out.

How hard would it be to dump the command line of the dotnet process you're about to kill, so we can see which one is the troublemaker?

4creators commented:
@AndyAyersMS 

> I'm not objecting to your change.

I understand that.

> How hard would it be to dump the command line of the dotnet process you're about to kill, so we can see which one is the troublemaker?

If there is any tool which we can use to create minidump from `dotnet` by invoking it from command line it would be a one liner.

4creators commented:
There is another possibility (known to me) with the external library `psutil` which can retrieve command line and many other properties directly.

This would require installation of `psutil` using `pip` and I don't know if this is feasible on CI machines. This lib is well known and widely used even by very large tech companies like Facebook and in general IMO it would be worth to use it instead of native python libs for system interaction.

AndyAyersMS commented:
On windows one can use powershell ... might require admin privs though
```
Get-WmiObject Win32_Process -Filter "name = 'dotnet.exe'" | Select-Object CommandLine
```
